akka-docs-dev/rst/scala/stream-composition.rst

```diff
@@ -0,0 +1,223 @@
+module itself, and (maybe somewhat surprisingly) it can be used as part of larger graphs. It is rarely useful to embed
+a closed graph shape in a larger graph (since it becomes an isolated island as there are no open port for communication
+with the rest of the graph), but this demonstrates the uniform underlying model.
+
+If we try to build a code snippet that corresponds to the above diagram, our first try might look like this:
+
+.. includecode:: code/docs/stream/CompositionDocSpec.scala#non-nested-flow
+
+It is clear however that there is no nesting present in our first attempt, since the library cannot figure out
+where we intended to put composite module boundaries, it is our responsibility to do that. If we are using the
+DSL provided by the :class:`Flow`, :class:`Source`, :class:`Sink` classes then nesting can be achieved by calling one of the
+methods ``withAttributes()`` or ``named()`` (where the latter is just a shorthand for adding a name attribute).
+
+The following code demonstrates how to achieve the desired nesting:
+
+.. includecode:: code/docs/stream/CompositionDocSpec.scala#nested-flow
+
+Once we have hidden the internals of our components, they act like any other built-in component of similar shape. If
+we hide some of the internals of our composites, the result looks just like if any other predefine component has been
+used:
+
+|
+
+.. image:: ../images/compose_nested_flow_opaque.png
+   :align: center
+
+|
+
+If we look at usage of built-in components, and our custom components, there is no difference in usage as the code
+snippet below demonstrates.
+
+.. includecode:: code/docs/stream/CompositionDocSpec.scala#reuse
+
+Composing complex systems
+-------------------------
+
+In the previous section we explored the possibility of composition, and hierarchy, but we stayed away from non-linear,
+generalized graph components. There is nothing in Akka Streams though that enforces that stream processing layouts
+can only be linear. The DSL for :class:`Source` and friends is optimized for creating such linear chains, as they are
+the most common in practice. There is a more advanced DSL for building complex graphs, that can be used if more
+flexibility is needed. We will see that the difference between the two DSLs is only on the surface: the concepts they
+operate on are uniform across all DSLs and fit together nicely.
+
+As a first example, let's look at a more complex layout:
+
+|
+
+.. image:: ../images/compose_graph.png
+   :align: center
+
+|
+
+The diagram shows a :class:`RunnableGraph` (remember, if there are no unwired ports, the graph is closed, and therefore
+can be materialized) that encapsulates a non-trivial stream processing network. It contains fan-in, fan-out stages,
+directed and non-directed cycles. The ``closed()`` method of the :class:`FlowGraph` object allows the creation of a
+general closed graph. For example the network on the diagram can be realized like this:
```

Comment:
` allows the creation of a general closed graph` + ` (and will throw at construction time if the constructed graph would be not fully connected)`?